### PR TITLE
Auto create clusters on login

### PR DESCRIPTION
**Pull request by MattPutnam**
Also added start/stop buttons so you can actually use your stuff.

### CODE REVIEW

File: src/components/AuthContainer.js [hunk 1]
```diff
@@ -38,6 +39,50 @@ export default class AuthContainer extends Component {
       }, () => {
         console.warn('Error looking up user status')
       })
+
+      Rawls.listBillingProjects().then(
+        billingProjects => {
+          this.billingProjects = billingProjects
+          this.checkClusters()
+        },
+        failure => Utils.log('Error fetching billing projects', failure)
```
Comment (dmohs) [inline src/components/AuthContainer.js]: TODO for me: Write a story to capture exposing errors to the user and logging them to New Relic.

File: src/components/AuthContainer.js [hunk 2]
```diff
@@ -38,6 +39,50 @@ export default class AuthContainer extends Component {
...
+    const { billingProjects, clusters } = this
+    if (billingProjects && clusters) {
+      let projectsWithoutClusters = _.difference(
+        _.map(billingProjects, 'projectName'),
+        _.map(clusters, 'googleProject')
+      )
+
+      if (projectsWithoutClusters.length > 20) {
+        console.log('More than 20 billing projects without clusters were found, only creating clusters for 20 of them')
+        projectsWithoutClusters = _.take(projectsWithoutClusters, 5)
```
Comment (dmohs) [inline src/components/AuthContainer.js]: Should this be 20? A constant that is used in each place?

File: src/pages/workspaces/workspace/Notebooks.js [hunk 3]
```diff
@@ -13,6 +13,44 @@ import * as Utils from 'src/libs/utils'
...
+  }
+
+  renderStopIcon() {
+    const { clusterName, googleProject, refresh } = this.props
+
+    return icon('pause',
+      {
+        class: 'is-solid', title: 'Stop Cluster',
+        style: { color: Style.colors.primary, cursor: 'pointer' },
+        onClick: () => Leo.cluster(googleProject, clusterName).stop().then(refresh)
```
Comment (dmohs) [inline src/pages/workspaces/workspace/Notebooks.js]: Ah! An obvious place where standard promise usage is better than async/await.

**Review by dmohs — APPROVED** [review]
Looks good to me. Is the intent that Brett's work will then remove the cluster list and move the status into the top navigation bar? I ask because this improved functionality makes it more obvious that when I'm viewing the cluster list, I'm seeing clusters that are outside of the current workspace's billing project.

File: src/components/AuthContainer.js [hunk 4]
```diff
@@ -38,6 +39,50 @@ export default class AuthContainer extends Component {
...
+          this.checkClusters()
+        },
+        failure => Utils.log('Error fetching clusters list', failure)
+      )
+    }
+  }
+
+  checkClusters = () => {
+    const { billingProjects, clusters } = this
+    if (billingProjects && clusters) {
```
Comment (panentheos) [inline src/components/AuthContainer.js]: In a case like this, where you have several independent async operations, you can use `Promise.all` to launch them all in parallel, and the master promise will resolve when they've all completed. So in this case you could do:
```
const [billingProjects, clusters] = await Promise.all([Rawls.listBillingProjects(), Leo.clustersList()])
// continue processing
```

File: src/pages/workspaces/workspace/Notebooks.js [hunk 5]
```diff
@@ -13,6 +13,44 @@ import * as Utils from 'src/libs/utils'
...
+  }
+
+  renderStartIcon() {
+    const { clusterName, googleProject, refresh } = this.props
+
+    return icon('play',
+      {
+        class: 'is-solid', title: 'Start Cluster',
+        style: { color: Style.colors.primary, cursor: 'pointer' },
+        onClick: () => Leo.cluster(googleProject, clusterName).start().then(refresh)
```
Comment (panentheos) [inline src/pages/workspaces/workspace/Notebooks.js]: TOL: It's generally good to have all clickable stuff be an `Interactive` when possible, so it gets a tabIndex and accessibility attributes.

**Review by panentheos — APPROVED** [review]
Looks good.

File: src/components/AuthContainer.js [hunk 6]
```diff
@@ -38,6 +39,27 @@ export default class AuthContainer extends Component {
       }, () => {
         console.warn('Error looking up user status')
       })
+
+      const [billingProjects, clusters] = await Promise.all([Rawls.listBillingProjects(), Leo.clustersList()])
+      if (billingProjects && clusters) {
```
Comment (panentheos) [inline src/components/AuthContainer.js]: nit: I don't think you need this check. If either call fails, you'll get an exception, so if the code gets here you can assume success. Looks good overall.

File: src/components/AuthContainer.js [hunk 7]
```diff
@@ -38,6 +39,25 @@ export default class AuthContainer extends Component {
       }, () => {
         console.warn('Error looking up user status')
       })
+
+      const [billingProjects, clusters] = await Promise.all([Rawls.listBillingProjects(), Leo.clustersList()])
```
Comment (zarsky-broad) [inline src/components/AuthContainer.js:43]: Is there a less convoluted way to do this without one blocking the other?

File: src/pages/workspaces/workspace/Notebooks.js [hunk 8]
```diff
@@ -13,6 +13,44 @@ import * as Utils from 'src/libs/utils'
...
+    ])
+  }
+
+  renderStartIcon() {
+    const { clusterName, googleProject, refresh } = this.props
+
+    return h(Interactive,
+      {
+        as: icon('play'), class: 'is-solid', title: 'Start Cluster',
+        style: { color: Style.colors.primary, cursor: 'pointer' },
```
Comment (zarsky-broad) [inline src/pages/workspaces/workspace/Notebooks.js]: I _believe_ you don't need to specify `cursor: 'pointer'` for an `Interactive` when there's an `onClick`.

File: src/components/AuthContainer.js [hunk 9]
```diff
@@ -38,6 +39,25 @@ export default class AuthContainer extends Component {
...
+      let projectsWithoutClusters = _.difference(
```
Comment (zarsky-broad) [inline src/components/AuthContainer.js:44]: Were we going to do something with not assuming that we can use clusters without a certain label?

**Review by zarsky-broad — APPROVED** [review]

File: src/components/AuthContainer.js (hunk 7 above)
Comment (panentheos) [reply src/components/AuthContainer.js:43]: [Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) is the standard mechanism for aggregating multiple promises. What seems convoluted about it?

Comment (zarsky-broad) [reply src/components/AuthContainer.js:43]: I think mainly the array destructuring from an array construction feels like a double negative to me.

File: src/components/AuthContainer.js (hunk 9 above)
Comment (dmohs) [reply src/components/AuthContainer.js:44]: That is tracked in #70.

**Pull request merged**